File: rack/drivers/servo_drive_proxy.cpp
/*
 * RACK - Robotics Application Construction Kit
 * Copyright (C) 2005-2010 University of Hannover
 *                         Institute for Systems Engineering - RTS
 *                         Professor Bernardo Wagner
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * Authors
 *      Joerg Langenberg  <joerg.langenberg@gmx.net>
 *
 */
#include <drivers/servo_drive_proxy.h>

int ServoDriveProxy::getData(servo_drive_data *recv_data, ssize_t recv_datalen,
                             rack_time_t timeStamp, uint64_t reply_timeout_ns)
{
    RackMessage msgInfo;

    int ret = RackDataProxy::getData((void *)recv_data, recv_datalen, timeStamp,
                                     reply_timeout_ns, &msgInfo);
    if (ret)
    {
        return ret;
    }

    recv_data = ServoDriveData::parse(&msgInfo);
    return 0;
}

int ServoDriveProxy::movePos(float position, float vel, float acc,
                             uint32_t replyPositionReached,
                             int32_t replyPositionReachedId,
                             uint64_t reply_timeout_ns)
{
    servo_drive_move_pos_data send_data;
    send_data.position             = position;
    send_data.vel                  = vel;
    send_data.acc                  = acc;
    send_data.replyPositionReached = replyPositionReached;

    return proxySendDataCmd(MSG_SERVO_DRIVE_MOVE_POS, &send_data,
                            sizeof(servo_drive_move_pos_data),
                            reply_timeout_ns);
}

int ServoDriveProxy::moveVel(float vel, uint64_t reply_timeout_ns)
{
    servo_drive_move_vel_data send_data;
    send_data.vel = vel;

    return proxySendDataCmd(MSG_SERVO_DRIVE_MOVE_VEL, &send_data,
                            sizeof(servo_drive_move_vel_data),
                            reply_timeout_ns);
}

File: rack/drivers/ServoDriveProxy.java
/*
 * RACK - Robotics Application Construction Kit
 * Copyright (C) 2005-2010 University of Hannover
 *                         Institute for Systems Engineering - RTS
 *                         Professor Bernardo Wagner
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 *
 * Authors
 *      Joerg Langenberg  <joerg.langenberg@gmx.net>
 *
 */
package rack.drivers;

import rack.main.*;
import rack.main.tims.*;
import rack.main.RackName;

public class ServoDriveProxy extends RackDataProxy
{
    public static final byte MSG_SERVO_DRIVE_HOME =
        RackProxy.MSG_POS_OFFSET + 1;

    public static final byte MSG_SERVO_DRIVE_MOVE_POS =
        RackProxy.MSG_POS_OFFSET + 2;

    public static final byte MSG_SERVO_DRIVE_MOVE_VEL =
        RackProxy.MSG_POS_OFFSET + 3;

    public static final byte MSG_SERVO_DRIVE_POSITION_REACHED =
        RackProxy.MSG_NEG_OFFSET - 1;

    public ServoDriveProxy(int system, int instance, TimsMbx replyMbx)
    {
        super(RackName.create(system, RackName.SERVO_DRIVE, instance, 0), replyMbx, 500);
    }

    public synchronized ServoDriveDataMsg getData(int recordingTime)
    {
        try
        {
            TimsRawMsg raw = getRawData(recordingTime);

            if (raw != null)
            {
                ServoDriveDataMsg data = new ServoDriveDataMsg(raw);
                return data;
            }
            else
            {
                return null;
            }
        }
        catch (TimsException e)
        {
            System.out.println(e.toString());
            return null;
        }
    }

    public synchronized ServoDriveDataMsg getData()
    {
        return (getData(0));
    }

    public synchronized void home()
    {
        try
        {

            replyMbx.send0(MSG_SERVO_DRIVE_HOME, commandMbx,
                           (byte) 0, (byte) 0);
            replyMbx.receive(0);
            System.out.println(RackName.nameString(replyMbx.getName()) + ": "
                    + RackName.nameString(commandMbx) + ".home");

        }
        catch (TimsException e)
        {
            System.out.println(RackName.nameString(replyMbx.getName()) + ": "
                    + RackName.nameString(commandMbx) + ".home" + e);
        }
    }

    public synchronized void movePos(float position, float vel, float acc,
            int replyPositionReached)
    {
        try
        {

            ServoDriveMovePosMsg command = new ServoDriveMovePosMsg();
            command.position = position;
            command.vel = vel;
            command.acc = acc;
            command.replyPositionReached = replyPositionReached;

            replyMbx.send(MSG_SERVO_DRIVE_MOVE_POS, commandMbx,
                          (byte) 0, (byte) 0, command);
            replyMbx.receive(0);

            System.out.println(RackName.nameString(replyMbx.getName()) + ": "
                    + RackName.nameString(commandMbx) + ".movePos position "
                    + position);
        }
        catch (TimsException e)
        {
            System.out.println(RackName.nameString(replyMbx.getName()) + ": "
                    + RackName.nameString(commandMbx) + ".movePos " + e);
        }
    }

    public synchronized void moveVel(float vel)
    {
        try
        {

            ServoDriveMoveVelMsg command = new ServoDriveMoveVelMsg();
            command.vel = vel;

            replyMbx.send(MSG_SERVO_DRIVE_MOVE_VEL, commandMbx,
                          (byte) 0, (byte) 0, command);
            replyMbx.receive(0);

            System.out.println(RackName.nameString(replyMbx.getName()) + ": "
                    + RackName.nameString(commandMbx) + ".moveVel vel " + vel);
        }
        catch (TimsException e)
        {
            System.out.println(RackName.nameString(replyMbx.getName()) + ": "
                    + RackName.nameString(commandMbx) + ".moveVel " + e);
        }
    }
}
